core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java

```diff
@@ -104,14 +104,17 @@ public TypeCoercionImpl(RelDataTypeFactory typeFactory, SqlValidator validator)
       updateInferredColumnType(scope1, query, columnIndex, targetType);
       return true;
     case VALUES:
+      boolean coercedValues = true;
       for (SqlNode rowConstructor : ((SqlCall) query).getOperandList()) {
         if (!coerceOperandType(scope, (SqlCall) rowConstructor, columnIndex, targetType)) {
-          return false;
+          coercedValues = false;
         }
       }
-      updateInferredColumnType(
-          requireNonNull(scope, "scope"), query, columnIndex, targetType);
-      return true;
+      if (coercedValues) {
+        updateInferredColumnType(
+            requireNonNull(scope, "scope"), query, columnIndex, targetType);
```

Comment:
What's the difference here ? The original code does the same behavior: `#updateInferredColumnType` only if all the operands are coerced.

Reply:
I thought so at first. But many test will be failed, for example:
```
 sql("select 1 from (values(true)) union values (select '1' from (values (true)) as tt)")
        .type("RecordType(VARCHAR EXPR$0) NOT NULL");
```
```
org.opentest4j.AssertionFailedError: expected: <RecordType(VARCHAR EXPR$0) NOT NULL> 
but was: <RecordType(VARCHAR NOT NULL EXPR$0) NOT NULL>
```
I think it is only executed when coerced values is true, like the following code
```      
// Update the nested SET operator node type.
if (coerced) {
  updateInferredColumnType(
      requireNonNull(scope, "scope"), query, columnIndex, targetType);
}
```

Reply:
> org.opentest4j.AssertionFailedError: expected: <RecordType(VARCHAR EXPR$0) NOT NULL> 
but was: <RecordType(VARCHAR NOT NULL EXPR$0) NOT NULL>

Shouldn't this be the correct behavior ? Two constant literals are type nullable false, and implicit type coercion does not solve nullability in-consistency.

Reply:
I modified the expected values of failed unit tests. They should all be not null. Are there any other questions?